File: Difficulty: Easy/Closest Neighbour in BST/closest-neighbour-in-bst.cpp
/*
class Node {
  public:
    int data;
    Node *left;
    Node *right;

    Node(int x) {
        data = x;
        left = NULL;
        right = NULL;
    }
}; */

class Solution {
  public:
    int findMaxFork(Node* root, int k) {
        int result = -1;  // Initialize result to -1 assuming no element found <= k

        while (root != NULL) {
            if (root->data <= k) {
                result = root->data;        // Current node is a valid candidate (<= k)
                root   = root->right;       // Try finding a bigger candidate on the right
            } else {
                root = root->left;          // Current node > k, go left to find smaller value
            }
        }

        return result; // Will hold the largest node value ≤ k, or -1 if no such value exists
    }
};

File: Difficulty: Easy/Closest Neighbour in BST/closest-neighbour-in-bst.java
/*
class Node {
    int data;
    Node left, right;

    Node(int x) {
        data = x;
        left = right = null;
    }
} */
class Solution {
    public int findMaxFork(Node root, int k) {
        // code here.
         int result = -1; // Initialize result with -1 in case no value ≤ k is found

        while (root != null) {
            if (root.data <= k) {
                result = root.data;     // Valid candidate, update result
                root   = root.right;    // Try to find a bigger candidate on the right
            } else {
                root = root.left;       // Too big, go left to find smaller value
            }
        }

        return result; // Return the closest value ≤ k
    }
}